File: packages/protocol/test/tamper.test.ts
// Adversarial cases the conformance gate requires: salt/field changes are caught
// (also exercised with fixture vectors in vectors.test.ts), duplicate keys
// and float-syntax numbers are rejected, and Merkle inclusion proofs with
// invalid padding/index/height are rejected.

import { describe, expect, it } from "vitest";
import {
  canonicalize,
  CanonicalError,
  frameHash,
  leafHash,
  nodeHash,
  MerkleTree,
  MerkleError,
  verifyInclusion,
} from "../src/index.js";

describe("strict canonicalization", () => {
  it("rejects a duplicate top-level key", () => {
    expect(() => canonicalize('{"a":1,"a":2}')).toThrow(CanonicalError);
  });

  it("rejects a duplicate nested key", () => {
    expect(() => canonicalize('{"outer":{"x":1,"x":2}}')).toThrow(CanonicalError);
  });

  it("rejects a duplicate key inside an array element", () => {
    expect(() => canonicalize('{"list":[{"x":1,"x":2}]}')).toThrow(CanonicalError);
  });

  it("rejects a float-syntax number", () => {
    expect(() => canonicalize('{"amount":1.5}')).toThrow(CanonicalError);
  });

  it("rejects an exponent-syntax number", () => {
    expect(() => canonicalize('{"amount":1e10}')).toThrow(CanonicalError);
  });

  it("rejects NaN/Infinity (not valid JSON syntax)", () => {
    expect(() => canonicalize('{"amount":NaN}')).toThrow(CanonicalError);
    expect(() => canonicalize('{"amount":Infinity}')).toThrow(CanonicalError);
  });

  it("accepts plain integers", () => {
    expect(() => canonicalize('{"count":42,"negative":-7,"zero":0}')).not.toThrow();
  });
});

describe("leaf/node hashing", () => {
  it("changes with the salt", () => {
    const record = { x: 1 };
    const a = leafHash(new Uint8Array(32).fill(1), record);
    const b = leafHash(new Uint8Array(32).fill(2), record);
    expect(a).not.toEqual(b);
  });

  it("changes with the field", () => {
    const salt = new Uint8Array(32).fill(1);
    const a = leafHash(salt, { x: 1 });
    const b = leafHash(salt, { x: 2 });
    expect(a).not.toEqual(b);
  });

  it("node hash is order sensitive", () => {
    const a = new Uint8Array(32).fill(1);
    const b = new Uint8Array(32).fill(2);
    expect(nodeHash(a, b)).not.toEqual(nodeHash(b, a));
  });
});

function sampleTree() {
  const leaves = Array.from({ length: 5 }, (_, i) => new Uint8Array(32).fill(i));
  return { tree: MerkleTree.build(leaves), leaves };
}

describe("merkle proof validation", () => {
  it("rejects an out-of-range prove() index", () => {
    const { tree } = sampleTree();
    expect(() => tree.prove(tree.count)).toThrow(MerkleError);
  });

  it("rejects verification with an out-of-range index", () => {
    const { tree, leaves } = sampleTree();
    const proof = tree.prove(0);
    proof.index = proof.count;
    expect(() => verifyInclusion(leaves[0]!, proof, tree.root)).toThrow(MerkleError);
  });

  it("rejects too few siblings (short padding)", () => {
    const { tree, leaves } = sampleTree();
    const proof = tree.prove(2);
    proof.siblings.pop();
    expect(() => verifyInclusion(leaves[2]!, proof, tree.root)).toThrow(MerkleError);
  });

  it("rejects excess siblings", () => {
    const { tree, leaves } = sampleTree();
    const proof = tree.prove(2);
    proof.siblings.push(new Uint8Array(32).fill(0xaa));
    expect(() => verifyInclusion(leaves[2]!, proof, tree.root)).toThrow(MerkleError);
  });

  it("rejects a tampered sibling value", () => {
    const { tree, leaves } = sampleTree();
    const proof = tree.prove(2);
    proof.siblings[0] = new Uint8Array(32).fill(0xff);
    expect(() => verifyInclusion(leaves[2]!, proof, tree.root)).toThrow(MerkleError);
  });

  it("rejects a proof claiming a different tree count", () => {
    const { tree, leaves } = sampleTree();
    const proof = tree.prove(2);
    proof.count = 6;
    expect(() => verifyInclusion(leaves[2]!, proof, tree.root)).toThrow(MerkleError);
  });

  it("has no valid proof for an empty tree", () => {
    const empty = MerkleTree.build([]);
    expect(() => empty.prove(0)).toThrow(MerkleError);
  });

  it("rejects a tampered leaf even with a correctly shaped proof", () => {
    const { tree } = sampleTree();
    const proof = tree.prove(2);
    const wrongLeaf = new Uint8Array(32).fill(0xee);
    expect(() => verifyInclusion(wrongLeaf, proof, tree.root)).toThrow(MerkleError);
  });
});

describe("framing", () => {
  it("succeeds for ordinary input", () => {
    expect(() => frameHash("tag", [new TextEncoder().encode("field")])).not.toThrow();
  });
});

File: crates/commitments/tests/tamper_test.rs
//! Adversarial cases the conformance gate requires: salt/field changes are caught
//! (also exercised with fixture vectors in vectors_test.rs), duplicate
//! keys and float-syntax numbers are rejected, and Merkle inclusion proofs
//! with invalid padding/index/height are rejected.

use commitments::*;
use serde_json::json;

#[test]
fn duplicate_top_level_key_is_rejected() {
    let result = canonicalize(r#"{"a":1,"a":2}"#);
    assert!(matches!(result, Err(CanonicalError::Parse(_))));
}

#[test]
fn duplicate_nested_key_is_rejected() {
    let result = canonicalize(r#"{"outer":{"x":1,"x":2}}"#);
    assert!(result.is_err());
}

#[test]
fn duplicate_key_inside_array_element_is_rejected() {
    let result = canonicalize(r#"{"list":[{"x":1,"x":2}]}"#);
    assert!(result.is_err());
}

#[test]
fn float_syntax_number_is_rejected() {
    let result = canonicalize(r#"{"amount":1.5}"#);
    assert!(result.is_err());
}

#[test]
fn exponent_syntax_number_is_rejected() {
    let result = canonicalize(r#"{"amount":1e10}"#);
    assert!(result.is_err());
}

#[test]
fn nan_and_infinity_are_not_valid_json_and_are_rejected() {
    // Not valid JSON syntax at all; the parser itself must reject these,
    // not silently coerce them.
    assert!(canonicalize(r#"{"amount":NaN}"#).is_err());
    assert!(canonicalize(r#"{"amount":Infinity}"#).is_err());
}

#[test]
fn integer_numbers_are_accepted() {
    assert!(canonicalize(r#"{"count":42,"negative":-7,"zero":0}"#).is_ok());
}

#[test]
fn changing_salt_changes_leaf_hash() {
    let record = json!({"x": 1});
    let a = leaf_hash(&[0x01; 32], &record).unwrap();
    let b = leaf_hash(&[0x02; 32], &record).unwrap();
    assert_ne!(a, b);
}

#[test]
fn changing_field_changes_leaf_hash() {
    let salt = [0x01; 32];
    let a = leaf_hash(&salt, &json!({"x": 1})).unwrap();
    let b = leaf_hash(&salt, &json!({"x": 2})).unwrap();
    assert_ne!(a, b);
}

#[test]
fn node_hash_is_order_sensitive() {
    let a = [0x01; 32];
    let b = [0x02; 32];
    assert_ne!(node_hash(&a, &b).unwrap(), node_hash(&b, &a).unwrap());
}

fn sample_tree() -> (MerkleTree, Vec<[u8; 32]>) {
    let leaves: Vec<[u8; 32]> = (0u8..5).map(|i| [i; 32]).collect();
    let tree = MerkleTree::build(leaves.clone()).unwrap();
    (tree, leaves)
}

#[test]
fn proof_with_index_out_of_range_is_rejected() {
    let (tree, _leaves) = sample_tree();
    let result = tree.prove(tree.count); // index == count is out of range
    assert!(matches!(result, Err(MerkleError::IndexOutOfRange { .. })));
}

#[test]
fn verification_rejects_index_out_of_range() {
    let (tree, leaves) = sample_tree();
    let mut proof = tree.prove(0).unwrap();
    proof.index = proof.count; // tamper: push index out of range
    let result = verify_inclusion(&leaves[0], &proof, &tree.root);
    assert!(matches!(result, Err(MerkleError::IndexOutOfRange { .. })));
}

#[test]
fn verification_rejects_too_few_siblings_short_padding() {
    let (tree, leaves) = sample_tree();
    let mut proof = tree.prove(2).unwrap();
    proof.siblings.pop(); // truncate: invalid padding, too short
    let result = verify_inclusion(&leaves[2], &proof, &tree.root);
    assert!(matches!(
        result,
        Err(MerkleError::InvalidSiblingCount { .. })
    ));
}

#[test]
fn verification_rejects_excess_siblings() {
    let (tree, leaves) = sample_tree();
    let mut proof = tree.prove(2).unwrap();
    proof.siblings.push([0xAA; 32]); // excess sibling: invalid padding
    let result = verify_inclusion(&leaves[2], &proof, &tree.root);
    assert!(matches!(
        result,
        Err(MerkleError::InvalidSiblingCount { .. })
    ));
}

#[test]
fn verification_rejects_wrong_sibling_value() {
    let (tree, leaves) = sample_tree();
    let mut proof = tree.prove(2).unwrap();
    proof.siblings[0] = [0xFF; 32]; // tamper a real sibling's value
    let result = verify_inclusion(&leaves[2], &proof, &tree.root);
    assert!(matches!(result, Err(MerkleError::RootMismatch)));
}

#[test]
fn verification_rejects_proof_for_different_tree_count() {
    let (tree, leaves) = sample_tree();
    let mut proof = tree.prove(2).unwrap();
    // Claim a different tree size while keeping the same siblings/index:
    // this must not be reinterpreted as a valid proof for a same-shaped
    // smaller/larger tree.
    proof.count = 6;
    let result = verify_inclusion(&leaves[2], &proof, &tree.root);
    assert!(result.is_err());
}

#[test]
fn empty_tree_has_no_valid_proof() {
    let empty = MerkleTree::build(vec![]).unwrap();
    assert!(matches!(empty.prove(0), Err(MerkleError::EmptyTree)));
}

#[test]
fn tampered_leaf_content_fails_verification_even_with_correct_proof_shape() {
    let (tree, _leaves) = sample_tree();
    let proof = tree.prove(2).unwrap();
    let wrong_leaf = [0xEE; 32]; // different leaf value, same proof
    let result = verify_inclusion(&wrong_leaf, &proof, &tree.root);
    assert!(matches!(result, Err(MerkleError::RootMismatch)));
}

#[test]
fn framing_overflow_is_a_typed_error_not_a_panic() {
    // Documents the overflow contract at the public API level; the
    // exhaustive boundary case lives in framing.rs's unit tests.
    let ok = frame_hash("tag", &[b"field"]);
    assert!(ok.is_ok());
}
